File: src/commonMain/kotlin/com/pricecompare/model/Money.kt
package com.pricecompare.model

/**
 * src/commonMain/kotlin/com/pricecompare/model/Money.kt
 *
 * 金額を表す値オブジェクト。JPY固定、小数点なし。
 * BigDecimalではなくDecimalを使用してKMP対応を実現。
 *
 * 関連ファイル:
 * - src/commonMain/kotlin/com/pricecompare/model/Decimal.kt
 * - src/commonMain/kotlin/com/pricecompare/pricing/PriceCalculator.kt
 * - src/commonMain/kotlin/com/pricecompare/model/PriceBreakdown.kt
 */
data class Money(val amount: Decimal) {

    operator fun plus(other: Money): Money = Money(amount + other.amount)
    operator fun minus(other: Money): Money = Money(amount - other.amount)
    operator fun times(factor: Decimal): Money = Money(amount * factor)
    operator fun div(other: Money): Decimal = amount / other.amount
    operator fun unaryMinus(): Money = Money(-amount)

    fun isPositive(): Boolean = amount.compareTo(Decimal.ZERO) > 0
    fun isNegative(): Boolean = amount.compareTo(Decimal.ZERO) < 0
    fun isZero(): Boolean = amount.compareTo(Decimal.ZERO) == 0

    fun abs(): Money = if (isNegative()) Money(-amount) else this

    override fun toString(): String = "${amount.toPlainString()}円"

    companion object {
        val ZERO = Money(Decimal.ZERO)

        fun of(value: Long): Money = Money(Decimal.fromLong(value))
        fun of(value: String): Money = Money(Decimal(value))
    }
}

operator fun Decimal.times(money: Money): Money = money * this

File: flutter_app/lib/engine/money.dart
import 'decimal.dart';

class Money implements Comparable<Money> {
  final Decimal amount;

  const Money(this.amount);

  static final Money zero = Money(Decimal.zero);

  factory Money.of(String value) => Money(Decimal(value));

  factory Money.fromInt(int value) => Money(Decimal.fromInt(value));

  Money operator +(Money other) => Money(amount + other.amount);

  Money operator -(Money other) => Money(amount - other.amount);

  Money operator *(Decimal factor) => Money(amount * factor);

  Decimal operator /(Money other) => amount / other.amount;

  Money operator -() => Money(-amount);

  bool get isPositive => amount.isPositive;
  bool get isNegative => amount.isNegative;
  bool get isZero => amount.isZero;

  Money abs() => isNegative ? -this : this;

  @override
  int compareTo(Money other) => amount.compareTo(other.amount);

  @override
  String toString() => '${amount.toPlainString()}円';

  @override
  bool operator ==(Object other) => other is Money && amount == other.amount;

  @override
  int get hashCode => amount.hashCode;
}
